### ISSUE TITLE
Unexpected "Object is possibly 'undefined'"

### ISSUE THREAD

**Issue by theodoretliu** [microsoft/TypeScript#39564]
**TypeScript Version:**  4.0.0-dev.20200711

**Search Terms:** Object is possibly 'undefined'

**Code**

```ts
const l = [
  {
    id: "id",
  },
];

export function demo1() {
  let x = l.find((_) => false);

  if (x !== undefined) {
    console.log(x.id); // no error on this line!
    l.find((e) => e.id === x.id); // but x is possibly undefined on this line!
  }
}

export function demo2(x: { id: string } | undefined) {
  if (x !== undefined) {
    console.log(x.id); // no error here
    l.find((e) => e.id === x.id); // AND no error here!
  }
}
```

run with `tsc --strictNullChecks code.ts`

**Expected behavior:** Expect `x.id` access in the `.find` of function `demo1` to be fine since there is the guard of `x !== undefined` and because `console.log(x.id)` throws no errors.

**Actual behavior:** Instead, within the anonymous function scope, the information about `x` has been erased and it has its original type of `{ id : string } | undefined`. Interestingly, in `demo2`, even though `x` has an identical type (or so I think), the information propagates to the anonymous function. This appears to be because `x` is a parameter in `demo2`.

**Playground Link:** https://repl.it/repls/LankyOnlyEmbeds 

**Related Issues:** https://github.com/microsoft/TypeScript/issues/8934

**Comment by MartinJohns** [comment]
Duplicate of #9998. Would be solved by #11498.

**Comment by theodoretliu** [comment]
Ah okay, so if I'm understanding properly, TS can't determine if at the time of the function invocation whether the local variable has been mutated back to `undefined` and so goes for the more relaxed analysis? @MartinJohns thanks for the quick response!
